Update closure and sharper closure minification

juj opened the pull request:
This is still WIP, but pushing as a PR to get the CI to run through it.

Investigating the reason why Closure does not minify as effectively as it could, the issue is that we are passing Closure suppression annotations using the Closure `--externs` option. Any annotations defined in `--externs` files have the effect that they also make Closure treat the symbol as something that it should not minify (in addition to applying whatever annotation was given to it).

To distinguish between annotating a symbol vs declaring a symbol as extern for Closure, this PR looks to split those into two different files, annotations are passed as a source file as `--js`, externs are passed as `--externs`.

Additionally, Closure externs are updated to be treated per target environments. If building code that is not supposed to be run in node.js or SpiderMonkey or V8, but only in browser, we should not declare symbols in those environments as extern, because Closure will not then minify them.

For example the two symbols `buffer` and `assert` that our preamble defines, are currently never minified because they are node.js externs. When building with `-s ENVIRONMENT='web'`, `buffer` and `assert` are not externs, so can and should be closure minified along with the rest of the symbols.

Not yet to merge, but looking for feedback and trying out a CI run to find remaining issues.

Had to update Closure to newer version, older version started giving internal compiler errors when I added multiple externs and js files.

kripken commented on tools/shared.py:
```diff
@@ -2448,23 +2460,36 @@ def closure_compiler(filename, pretty=True):
+        for extern in NODE_EXTERNS:
+          args.append('--externs')
+          args.append(extern)
+      if Settings.target_environment_may_be('web'):
```
this should probably be `web or worker`

kripken commented on tools/shared.py:
```diff
@@ -2448,23 +2460,36 @@ def closure_compiler(filename, pretty=True):
+        for extern in BROWSER_EXTERNS:
+          args.append('--externs')
+          args.append(extern)
+      if Settings.target_environment_may_be('v8'):
```
there is `shell` that means `v8 or spidermonkey or jsc`

juj commented:
Updated the PR, this is now good for review and landing.

The following Google Closure bug tracker issues were created related to this:
 - https://github.com/google/closure-compiler/issues/3191 : The use of `--js=closure-annotations.js` as a mechanism to get annotated JS symbols to minify has a limitation in Closure that these will in some cases create redundant minified `var k;` directives at the top of the generated output source file. This is not ideal and causes bloat, but still better than foregoing minification of those symbols in question altogether.
 - https://github.com/google/closure-compiler/issues/3189 : New version of Closure has a regression that it no longer allows functions to be defined inside a block scope `{` `}`, if `--language_in=ES5` is used. Iiuc it no longer has support for treating functions with function scope, but now always treats functions with block scope. As a workaround, changed such functions from `function foo(){}` to `var foo = function(){}`. This introduced limitation in Closure can affect code in users' own JS libraries. If this is an issue, I can try reverting Closure back to the older version.
 - https://github.com/google/closure-compiler/issues/3187 : a pattern that does not currently optimize out from our codegen (we have had this always)
 - https://github.com/google/closure-compiler/issues/3185 : another pattern that does not optimize away (also has been present prior to this PR)

juj commented on src/closure-annotations.js:
```diff
@@ -0,0 +1,325 @@
+/**
+* @type {Object.<string,*>}
+*/
+var flags = {};
```
I don't know, I retained it from the original Closure externs file. I think followup PRs can investigate to remove more things that seem redundant.

juj commented on src/closure-annotations.js:
```diff
@@ -0,0 +1,325 @@
+var WasmJS;
+
+// Various variables that get into WasmJS of Binaryen, but are not present in final build and never actually used (should probably be fixed in future)
+// These are only needed for `interpret-asm2wasm` and not for `asmjs` or `native-wasm`
```
I will leave this for a later optimization, removing these did not seem to be safe, and I don't have time to investigate this further.

juj commented:
Comparing code sizes after update, noticed that unfortunately new Closure compiler actually regresses compiled code size. I was able to avoid the internal compiler error that I was getting before, so reverted back to old Closure version.

I was able to narrow down one pattern that regresses, reported at https://github.com/google/closure-compiler/issues/3193 .

The name of this PR branch is still "update_closure_...", but this PR no longer changes the closure version, reverted it back to the old version (also due to the other Closure issue 3189).

Code size wise, minimal_hello.c goes from 8281 bytes to 8179 bytes (-1.23%) and webgl_draw_triangle.c goes from 15754 bytes to 15284 bytes (-2.98%).

@kripken: There are a couple of new changes to JS libraries and preamble since your previous review to help reduce the number of annotations that we need, to sidestep issue https://github.com/google/closure-compiler/issues/3191.

kripken commented on src/library_html5.js:
```diff
@@ -1670,14 +1671,14 @@ var LibraryJSEvents = {
     }
 
-    return JSEvents.requestFullscreen(target, strategy);
+    return _JSEvents_requestFullscreen(target, strategy);
```
why not make this depend on `$JSEvents`? is it to save code size I guess?

how about doing `$JSEvents_requestFullscreen` for readability, like other helper functions?

juj replied:
The intent is for this symbol to be private, so do not want to `$` that one. These needed to be pulled away from JSEvents so that JS linker would be able to track these without pulling the whole JSEvents object in.

kripken commented on src/preamble.js:
```diff
@@ -1830,7 +1830,7 @@ function integrateWasmJS() {
   // Create the wasm instance.
   // Receives the wasm imports, returns the exports.
-  function createWasm(global, env, providedBuffer) {
+  function createWasm(env) {
```
oh nice, these were leftovers from the binaryen asm.js support we removed, good to remove them

kripken commented on tools/shared.py:
```diff
@@ -1239,6 +1239,12 @@ def get(self, key):
+    @classmethod
+    def target_environment_may_be(self, environment):
+      SHELL_ENVIRONMENTS = ['spidermonkey', 'jsc', 'v8']
+      targeted_environments = self.attrs['ENVIRONMENT'].split(',')
```
we don't intend for users to write `ENVIRONMENT=jsc` etc. - all we support there is `shell,web,worker,node`.

seems fine to me to keep the closure stuff in separate files, but it would be confusing to support `ENVIRONMENT=v8` etc. here, please remove it

kripken replied:
to be clearer, what I'm suggesting is that the closure method not call this with `v8`, `jsc`, etc., since it implies those are valid environments. Instead, how about if it calls this with `shell`, and in the closure method when that happens, it includes the `v8`, `jsc` etc. files?

kripken replied:
(as this is a detail of how the closure support is implemented, but those are not valid ENVIRONMENTs as that is defined in src/settings.js)

juj replied:
Oh gotcha. Updated.

kripken commented:
Nice code size win here!

juj commented:
Oh hmm, turns out that I do need to update Closure after all to make all of this work, so restored back newer Closure version to the build. Hopefully all tests now pass for this to be good to merge.

juj commented:
The newly added test `browser.test_closure_in_web_only_target_environment` does seem to consistently fail to a timeout in the Chrome circleci suite, but when I run the test locally on macOS or Windows in Chrome, the test passes. I don't know how to proceed. Trying to split up the test into two parts, in case it takes too long to execute or something.

kripken commented:
I don't think it's a time out - the `test_closure_in_web_only_target_environment_webgl` test needs an `@requires_graphics_hardware` annotation (which would then make it be skipped on the chrome bot).

kripken commented:
(What is the overall code size change of this PR, after restoring the new closure to here?)

juj commented:
Thanks, added `@requires_graphics_hardware`.

The code size is about the same as before the closure update.

juj commented:
The remaining failures look like something with the CI perhaps? The tests do pass for me locally on my Windows and Linux boxes.

kripken commented:
CI is now working again - may need to push a new commit here to get it running though.

juj commented:
This PR shows the same CI failures as #7863, although there is also

```
FAIL: test_stack_varargs2 (test_core.wasmobj2)
```

but I don't have a setup currently where I could try that out. How does this look to you?

juj commented:
Oh hmm, #7834 also shows failure on FAIL: test_stack_varargs2 (test_core.wasmobj2). Perhaps that is unrelated to this PR?

kripken commented:
Can ignore `test_stack_varargs2 (test_core.wasmobj2)`, I see that in other PRs too. Trying to fix in #7893 

I'm not sure about the single_file errors, which I've also seen elsewhere. I reran the job now.

kripken commented:
The `single_file` errors are a firefox nightly change, can ignore them too.
